pkg/provider/velero/backupProvider.go

```diff
@@ -0,0 +1,204 @@
+	factory := veleroclient.NewFactory("k8s-backup-operator", map[string]interface{}{"namespace": namespace})
+	clientSet, err := factory.Client()
+	if err != nil {
+		return nil, fmt.Errorf("failed to create velero clientset: %w", err)
+	}
+
+	return &provider{recorder: recorder, veleroClientSet: clientSet, namespace: namespace}, nil
+}
+
+// CheckReady validates that velero is installed and can establish a connection to its backup store.
+func (p *provider) CheckReady(ctx context.Context) error {
+	defaultBsl, err := p.veleroClientSet.VeleroV1().BackupStorageLocations(p.namespace).Get(ctx, "default", metav1.GetOptions{})
+	if err != nil {
+		return fmt.Errorf("failed to get backup storage location from cluster: %w", err)
+	}
+
+	if defaultBsl.Status.Phase != velerov1.BackupStorageLocationPhaseAvailable {
+		return fmt.Errorf("velero is unable to reach the default backup storage location: %s", defaultBsl.Status.Message)
+	}
+
+	return nil
+}
+
+// CreateBackup triggers a velero backup and waits for its completion.
+func (p *provider) CreateBackup(ctx context.Context, backup *v1.Backup) error {
+	p.recorder.Event(backup, corev1.EventTypeNormal, v1.CreateEventReason, "Using velero as backup provider")
+
+	volumeFsBackup := false
+	veleroBackup := &velerov1.Backup{
+		ObjectMeta: metav1.ObjectMeta{Name: backup.Name, Namespace: backup.Namespace},
+		Spec: velerov1.BackupSpec{
+			IncludedNamespaces:       []string{backup.Namespace},
+			StorageLocation:          "default",
+			DefaultVolumesToFsBackup: &volumeFsBackup,
+		},
+	}
+
+	veleroBackupClient := p.veleroClientSet.VeleroV1().Backups(backup.Namespace)
+	_, err := veleroBackupClient.Create(ctx, veleroBackup, metav1.CreateOptions{})
+	if err != nil {
+		return p.handleFailedBackup(backup, fmt.Errorf("failed to apply velero backup '%s/%s' to cluster: %w", veleroBackup.Namespace, veleroBackup.Name, err))
+	}
+
+	watcher, err := veleroBackupClient.Watch(ctx, metav1.ListOptions{FieldSelector: backup.GetFieldSelectorWithName()})
+	if err != nil {
+		return p.handleFailedBackup(backup, fmt.Errorf("failed to create watch for velero backup '%s/%s': %w", veleroBackup.Namespace, veleroBackup.Name, err))
+	}
+
+	veleroBackupChan := watcher.ResultChan()
+	defer watcher.Stop()
+
+	err = waitForBackupCompletionOrFailure(veleroBackupChan)
+	if err != nil {
+		return p.handleFailedBackup(backup, err)
+	}
+
+	p.recorder.Eventf(backup, corev1.EventTypeNormal, v1.CreateEventReason, "Successfully completed velero backup '%s/%s'", veleroBackup.Namespace, veleroBackup.Name)
+	return nil
+}
+
+func (p *provider) handleFailedBackup(backup *v1.Backup, err error) error {
+	p.recorder.Event(backup, corev1.EventTypeWarning, v1.ErrorOnCreateEventReason, err.Error())
+	return err
+}
+
+func waitForBackupCompletionOrFailure(veleroBackupChan <-chan watch.Event) error {
+eventLoop:
+	for veleroChange := range veleroBackupChan {
+		changedBackup := veleroChange.Object.(*velerov1.Backup)
```

Comment:
🚫 **[checkstyle]** <forcetypeassert> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>type assertion must be checked